Rubberduck.Inspections/Concrete/ShadowedDeclarationInspection.cs

```diff
@@ -0,0 +1,232 @@
+                                declaration.IdentifierName,
+                                RubberduckUI.ResourceManager.GetString("DeclarationType_" + shadowedDeclaration.DeclarationType,CultureInfo.CurrentUICulture),
+                                shadowedDeclaration.IdentifierName),
+                            declaration));
+                    }
+                } 
+            }
+
+            return issues;
+        }
+
+        private static bool DeclarationIsPartOfBuiltInEventHandler(Declaration declaration, ICollection<Declaration> builtInEventHandlers)
+        {
+            if (builtInEventHandlers.Contains(declaration))
+            {
+                return true;
+            }
+
+            var parameterDeclaration = declaration as ParameterDeclaration;
+
+            return parameterDeclaration != null && builtInEventHandlers.Contains(parameterDeclaration.ParentDeclaration);
+        }
+
+        private bool DeclarationCanBeShadowed(Declaration originalDeclaration, Declaration userDeclaration)
+        {
+            var originalDeclarationComponentType = originalDeclaration.QualifiedName.QualifiedModuleName.ComponentType;
+            var userDeclarationComponentType = userDeclaration.QualifiedName.QualifiedModuleName.ComponentType;
+
+            // It is not possible to directly access a Parameter, UDT Member or Label declared in another project
+            if (originalDeclaration.DeclarationType == DeclarationType.Parameter || originalDeclaration.DeclarationType == DeclarationType.UserDefinedTypeMember || 
+                originalDeclaration.DeclarationType == DeclarationType.LineLabel)
+            {
+                return false;
+            }
+
+            // It is not possible to instantiate a Class Module which is not exposed
+            if ((originalDeclaration as ClassModuleDeclaration)?.IsExposed == false)
+            {
+               return false;
+            }
+
+            // It is not possible to directly access a UserForm or Document declared in another project, nor any declarations placed inside them
+            if (originalDeclarationComponentType == ComponentType.UserForm || originalDeclarationComponentType == ComponentType.Document)
+            {
+                return false;
+            }
+
+            // It is not possible to directly access any declarations placed inside a Class Module
+            if (originalDeclaration.DeclarationType != DeclarationType.ClassModule && originalDeclarationComponentType == ComponentType.ClassModule)
+            {
+                return false;
+            }
+
+            if (userDeclaration.DeclarationType == DeclarationType.ClassModule)
+            {
+                if (userDeclarationComponentType == ComponentType.UserForm && !TypeShadowingRelations[originalDeclaration.DeclarationType].Contains(DeclarationType.UserForm))
```

Comment:
wants to be shortened to `!TypeShadowingRelations[originalDeclaration.DeclarationType].Contains(userDeclarationComponentType)`, which also collapses the next if-block into this. 

That would need to allow getting a ComponentType to pass as DeclarationType ... not sure whether that's useful elsewhere, too?